src/projection.jl

```diff
@@ -31,114 +36,317 @@ function Base.show(io::IO, project::ProjectTo{T}) where {T}
+    element = ProjectTo(zero(T))
+    # if all our elements are going to zero, then we can short circuit and just send the whole thing
+    element isa ProjectTo{<:AbstractZero} && return element
+    return ProjectTo{AbstractArray}(; element=element, axes=axes(x))
 end
-function (project::ProjectTo{T})(dx::AbstractZero) where {T<:Array}
-    return T(map(proj -> proj(dx), project.elements))
+function (project::ProjectTo{AbstractArray})(dx::AbstractArray{S,M}) where {S,M}
+    T = project_type(project.element)
+    dy = S <: T ? dx : map(project.element, dx)
+    if axes(dy) == project.axes
+        return dy
+    else
+        # The rule here is that we reshape to add or remove trivial dimensions like dx = ones(4,1),
+        # where x = ones(4), but throw an error on dx = ones(1,4) etc.
+        for d in 1:max(M, length(project.axes))
+            size(dy, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dy)))
+        end
+        return reshape(dy, project.axes)
+    end
 end
-(project::ProjectTo{<:Array})(dx::AbstractArray) = project(collect(dx))
 
-# Arrays{<:Number}: optimized case so we don't need a projector per element
-function ProjectTo(x::T) where {E<:Number,T<:Array{E}}
-    return ProjectTo{T}(; element=ProjectTo(zero(E)), size=size(x))
+# Zero-dimensional arrays -- these have a habit of going missing:
+function (project::ProjectTo{AbstractArray})(dx::Number) # ... so we restore from numbers
+    project.axes isa Tuple{} || sum(length, project.axes) == 1 || throw(_projection_mismatch(project.axes, size(dx)))
+    fill(project.element(dx))
 end
-(project::ProjectTo{<:Array{T}})(dx::Array) where {T<:Number} = project.element.(dx)
-function (project::ProjectTo{<:Array{T}})(dx::AbstractZero) where {T<:Number}
-    return zeros(T, project.size)
+
+# Arrays of arrays -- store projector per element:
+ProjectTo(xs::AbstractArray{<:AbstractArray}) = ProjectTo{AbstractArray{AbstractArray}}(; elements=map(ProjectTo, xs), axes = axes(xs))
+function (project::ProjectTo{AbstractArray{AbstractArray}})(dx::AbstractArray)
+    dy = if axes(dx) == project.axes
+        dx
+    else
+        for d in 1:max(ndims(dx), length(project.axes))
+            size(dx, d) == length(get(project.axes, d, 1)) || throw(_projection_mismatch(project.axes, size(dx)))
+        end
+        reshape(dx, project.axes)
+    end
+    # This always re-constructs the outer array, it's not super-lightweight
+    return map((f,x) -> f(x), project.elements, dy)
 end
-function (project::ProjectTo{<:Array{T}})(dx::Tangent{<:SubArray}) where {T<:Number}
-    return project(dx.parent)
+
+# Arrays of other things -- since we've said we support arrays, but may not support their elements,
+# we handle the container as above but store trivial element projector:
+ProjectTo(xs::AbstractArray) = ProjectTo{AbstractArray}(; element=ProjectTo(), axes=axes(xs))
+
+# Ref -- likewise aim at containers of supported things, but treat unsupported trivially.
```

Comment:
I am not sold on this aim.
Do we have a convincing use case where someone is going to do something unsupported?
And it is better to just kinda guess than error and make them fix it.
Maybe something with Unitful?

I can believe this for AbstractArray subtypes, where we just guess they are dense.
but for arbitary objects I am much less sold.
(and I was dubious for AbstractArrays)

Reply:
Zygote has a habit of making `Ref{Any}`. Like `Array{Any}`, I feel that leaving making these produce an error where the typed one would not is pretty surprising. 

It's possible that both `Ref{Any}` and `Array{Any}` should be handled by looking at the contents (and ignoring it if it's strings) instead of just giving up.

Reply:
Now Ref{Any}(3) will work exactly like Ref(3), but Any[1,2,3] will still not look at the contents, it will only note that it is a vector of length 3.